starship: add enable_transience for fish

will opened the pull request:
### Description

Starship has an advanced, experimental feature where fancy stuff in the prompt can be replaced with something more simple after the command is ran. This is very helpful for copy and pasting shell history somewhere else.

<img width="682" alt="image" src="https://github.com/nix-community/home-manager/assets/1973/a45c2949-7800-45e5-bf0d-6c98555641d8">

docs: https://starship.rs/advanced-config/#transientprompt-and-transientrightprompt-in-fish

Fish is currently the only shell as far as I can tell that both home-manager and starship support for this feature. Since the function has to be called after starship is loaded, this seems like the best place to put it. Without putting it inside home-manager itself, you would have to forego the provided `programs.fish.interactiveShellInit` entirely just so you can be sure to only call the new function after starship is loaded, which seems like a shame.


### Checklist

<!--

Please go through the following checklist before opening a non-WIP
pull-request.

Also make sure to read the guidelines found at

  https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-guidelines

-->

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [ ] Code tested through `nix-shell --pure tests -A run.all`.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

I did run just the starship-settings test, but there weren't any tests for the shell init hooks.

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

ncfavier commented on modules/programs/starship.nix:
```diff
@@ -101,6 +103,7 @@ in {
     programs.fish.interactiveShellInit = mkIf cfg.enableFishIntegration ''
       if test "$TERM" != "dumb"  -a \( -z "$INSIDE_EMACS"  -o "$INSIDE_EMACS" = "vterm" \)
         eval (${starshipCmd} init fish)
+        ${lib.optionals cfg.enableTransience "enable_transience"}
```
Have you tested this? I'd expect it to fail since `optionals` produces a list and you're interpolating a string...

Please add a test

will replied:
> Have you tested this? 

Yes, I'm running currently off of my branch, and the screenshot in the PR description is from using it. The bottom half where it has just `>` instead of the colors is the feature. 

```nix
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
    stable.url = "github:NixOS/nixpkgs/22.11";
    #home-manager = { url = "github:nix-community/home-manager"; inputs.nixpkgs.follows = "nixpkgs"; };
    home-manager = { url = "github:will/home-manager/starship-transience"; inputs.nixpkgs.follows = "nixpkgs"; };
```

However I'll switch the conditional to a more suitable one.

> Please add a test

I'm all for adding tests, but haven't yet written any tests in nix. And the other related tests for starship only have tests for the generated toml file, and not for the hooks it puts into shells. Do you have an example of another project that has a test for the `interactiveShellInit` hooks that I can learn from?

ncfavier replied:
Right, it only fails when `enableTransience` is `false` since `optionals false ...` yields `[]`. Use `optionalString` instead.

ncfavier commented on modules/programs/starship.nix:
```diff
@@ -77,6 +77,8 @@ in {
       default = true;
     };
+
+    enableTransience = mkEnableOption "TransientPrompt" // { default = false; };
```
This should link to the docs and make it clear that only fish supports this.

will commented:
Thanks for the review @ncfavier. For updates to the PR, do you prefer new fixup commits to this, or force pushing over the existing commit, or something else?

will commented:
Ok @ncfavier, I've addressed both of your feedback. I've used `optionalString` properly and added tests for both fish with and without transience enabled, and I've added documentation that it's fish only with the url linking to the starship docs. 

I think that's everything, please let me know if there are other issues though
